BUG/MAINT: Change default of inplace to False in pd.eval

gfyoung opened the pull request:
Deprecated back in 0.18.0.

xref #11149.

Also patches bug where we were improperly handling the `inplace=False` condition, as we were assuming that target input was non-None when that wasn't necessarily enforced.

gfyoung commented:
Apparently > 200 (!) of our tests didn't take heed of the warning to pass in `inplace=True`

gfyoung commented:
Also, I must say that the docstring for `inplace` is not very intuitive IMO.  I would have expected that `inplace` would return the result of the computation and not the other way around.

jorisvandenbossche commented on pandas/core/computation/eval.py:
```diff
@@ -207,13 +206,10 @@ def eval(expr, parser='pandas', engine=None, truediv=True,
-        in a future version, will default to False.  Use inplace=True
-        explicitly rather than relying on the default.
+    inplace : bool, default False
+        If the expression mutates, whether to modify the expression inplace
```
I think adding an example might clarify it further eg: "If the expression mutates (e.g. creates a new column), ..."

jorisvandenbossche replied:
You changed object to expression in "whether to modify object inplace", but I think object is actually better.

Looking at the test example:
```
In [87]: df = DataFrame(np.random.randn(5, 2), columns=list('ab'))

In [88]: df.eval('c = a + b', inplace=False)
Out[88]: 
          a         b         c
0 -0.742101  2.084071  1.341970
1 -0.823597 -1.686791 -2.510388
2  0.277927 -0.076424  0.201503
3  1.159483  0.188500  1.347983
4  0.010328 -0.890899 -0.880571

In [89]: df
Out[89]: 
          a         b
0 -0.742101  2.084071
1 -0.823597 -1.686791
2  0.277927 -0.076424
3  1.159483  0.188500
4  0.010328 -0.890899
```

it is the `df` object that does or not get modified, not the "expression"

jorisvandenbossche commented on pandas/core/computation/eval.py:
```diff
@@ -207,13 +206,10 @@ def eval(expr, parser='pandas', engine=None, truediv=True,
+    inplace : bool, default False
+        If the expression mutates, whether to modify the expression inplace
+        or return a copy of it with mutation. In other words, it will return
+        the result of the expression if `inplace=True` and `target` otherwise.
```
"and `target` otherwise." is not fully correct. It can also mutate the calling object.

WIth the same example:

```
In [89]: df
Out[89]: 
          a         b
0 -0.742101  2.084071
1 -0.823597 -1.686791
2  0.277927 -0.076424
3  1.159483  0.188500
4  0.010328 -0.890899

In [90]: df.eval('c = a + b', inplace=True)

In [91]: df
Out[91]: 
          a         b         c
0 -0.742101  2.084071  1.341970
1 -0.823597 -1.686791 -2.510388
2  0.277927 -0.076424  0.201503
3  1.159483  0.188500  1.347983
4  0.010328 -0.890899 -0.880571
```

jorisvandenbossche reviewed (commented):
Adding an example to the docstring might clarify a lot what this function actually does

jorisvandenbossche commented:
Whoops, sorry, all my comments were in light of me thinking it was about `DataFrame.eval` method, not top-level `pandas.eval`. 

Anyhow, the frame method also has to be updated, and there the comments will be sensible :-)

gfyoung commented:
@jorisvandenbossche : I agree that `inplace` makes a lot more sense in the context of `DataFrame`, but in the context of `pd.eval`, it lacks the same sensibility.  A lot of the (failing) tests pass in `target=None`, so what in the world does `inplace=False` mean then?

And if it's truly inplace, why are we returning anything then when `inplace=True`?  That's inconsistent with what I would expect an inplace-operation to do.

IMO this is an API issue because `inplace` doesn't make much sense without a clearly defined target in the first place, and the tests don't do much to help there.

gfyoung commented:
@jorisvandenbossche : Actually ignore my previous comment.  That doesn't make much sense to me at all still.  The documentation says "whether to return a new DataFrame or mutate the existing."  Which condition means return a new `DataFrame`?

If you say it's `inplace=False`, then why are we returning `target` in `pd.eval` (and not `ret`)?  Also, why are we returning anything when `inplace=True` in `pd.eval`?  IMO, it should be that `ret` is returned when `inplace=False`, and nothing is returned when `inplace=True` because `target` should be modified.  In fact, I might go as far as to say that `inplace=True` only when `target` is not a primitive.  Not really sure what this means here:

~~~python
>>> import pandas as pd
>>> c = pd.eval('1 + 2', inplace=True)
>>> c
3
>>> c = pd.eval('1 + 2', inplace=False)
>>> c
>>>
~~~

jorisvandenbossche commented:
cc @chris-b1

chris-b1 commented:
My thought is that `inplace` should only be part of the api for `DataFrame.eval` - there it works as expected, and consistent with other uses of `inplace.  For `pd.eval` it is simply an implementation detail (since `DataFrame.eval` calls it) - I should not have added it to the docstring.

In fact, `pd.eval` should probably raise if `inplace` is passed in (so keep the default = None) without a `target`.

gfyoung commented:
@chris-b1 : I think we are mostly aligned in our thought process, and I think (if nobody objects) that we should address the issue now in this PR instead of waiting.

I don't think adding `inplace` to the doc-string was a bad idea.  It's just that I don't understand why it behaves so counter-intuitively to what I would think `inplace` does in `pd.eval`.  Why is it that `target` is returned when `inplace=False`?  It should be the other way around.

I agree that we should only allow `inplace=True` when `target` is not a primitive.  Furthermore, we should be returning anything when `inplace=True` because we assume that `target` itself has been modified.

What do you think?

jorisvandenbossche commented:
The question we should ask: what is the usecase of `inplace`? (is there actually a usecase?) 
I don't see one, so I agree with @chris-b1 we better remove it from the docstring / document it as no effect.

Possible usecases:

* expression without assignment:
  * this should always return something, as there is no object to return to (passing a target also, logically, does not do anything with the target). So `inplace` makes no sense in this case.
  * The reason that `inplace=False` actually suppresses the return value is just a bug in the current implementation. The code does:

    ```
    if not inplace and inplace is not None:
        return target

    return ret
    ```

    which fails for inplace=False without a target.

* expression with assignment:
  * for this case, you are obliged to pass a `target`, so why would you not want to modify this target?
  * That's my first thought, but OK, here might indeed be a possible usecase: eg you want to add a column to an existing dataframe, but still want to take a copy of it first, and then return the modified result instead of modifying inplace. Similar as we do with eg `assign`

So actions I would take:

- Fix the bug with inplace in case of no assignment. And I would even additionally raise an error if `inplace` is specified in such a case
- Update the docstring to make it clear that `inplace` is only relevant for the case where a `target` is passed.

gfyoung commented:
@jorisvandenbossche : Well that touches upon #16529, but in any case, it can be used when `target` is a `DataFrame` object, so I wouldn't say that it has no use case at this point.

I agree that we should update the docstring to say that `inplace` has no meaning when `target` isn't a non-primitive object (passing in `target=1` also makes no sense).

Also, the fix for that bug is the following:
~~~python
<if target is None or is some primitive>:
   inplace = False
...
if not inplace:
   return ret
~~~

There should be no condition for returning if `inplace=True` because we assume that `target` itself is modified.

chris-b1 commented:
I would suggest

 - DataFrame.eval`, change `inplace` default to `False`
 - `pd.eval`

    - keep default for `inplace` as `None`
    - if `inplace` kw is passed with no target, raise

gfyoung commented:
@chris-b1 : I agree with `inplace=False`, but I don't agree with your suggestion about `inplace=None`.  The only time I would consider raising is when `inplace=True` (`inplace=False` shouldn't care about `target`) and `target` is something that you can't modify in-place.  Also, I think we should default `inplace=False` to `pd.eval`, as that is consistent with what we have for other `inplace` defaults.

jorisvandenbossche commented on pandas/core/computation/eval.py:
```diff
@@ -232,8 +237,23 @@ def eval(expr, parser='pandas', engine=None, truediv=True,
+    # "Primitives" cannot be modified inplace because
+    # they have no attributes that we can modify.
+    #
+    # The only exceptions are lists and dictionaries,
```
I don't think it is possible to modify lists -> `__getitem__` is used, and it is a string expression, so you can never work with a list

Or do you have an example where this actually works?

(anyway, it would also be good to add some tests for those statements, also for the dictionaries)

jorisvandenbossche replied:
Ah, I see the diff I was looking at is already outdated :-)

gfyoung replied:
@jorisvandenbossche : This is very much WIP.  I'm trying to test that the changes I make actually don't break anything, let alone > 200 tests like last time. :smile:

gfyoung replied:
Once I can rewrite the implementation in a way that works, I will add the tests (I have them in a separate file that I'm running to test before pushing).

jorisvandenbossche commented on pandas/core/computation/eval.py:
```diff
@@ -232,8 +237,13 @@ def eval(expr, parser='pandas', engine=None, truediv=True,
+    modifiable = hasattr(target, "__setitem__")
+
+    if inplace and not modifiable:
+        raise ValueError("Cannot modify the provided target inplace")
```
Another possibility is to just raise depending on `parsed_expr.assigner`. If there is no assignment in the expression, raise when inplace is set to True

jorisvandenbossche replied:
Because now you could pass a target, and then set inplace=True (then this error will not be triggered), while this makes no sense when there is no assignment

gfyoung replied:
> Because now you could pass a target, and then set inplace=True

What do you mean?  How can you do those two things sequentially?

jorisvandenbossche replied:
Just using both as keyword arguments: pd.eval(expr, target=df, inplace=True)

that only makes sense when expr contains an assignment, and that you check with `parsed_expr.assigner`

gfyoung replied:
That's fair.  I think I'll just put a `try-except` around the item assignment so that we don't have to do all of these checks along the way.

jorisvandenbossche replied:
For me the try excepts are not needed. When trying to set an object that does not support assigment with strings like a list, you already get a clear error message (and this is really a corner case, people should actually only pass dataframes)

gfyoung replied:
I think a uniform error message is needed.  Passing in `np.ndarray` provides an error message that is not immediately clear regarding item assignment.

> people should actually only pass dataframes

Well perhaps `pd.eval` should not exist. :smile:  Because as is, there's nothing wrong with me passing in a dictionary or some other dict-like object.

gfyoung commented:
@jreback @jorisvandenbossche : Could you run the failing build on Travis?  It doesn't seem related to my changes at all whatsoever.

gfyoung commented:
By the way, this PR is ready for review, as I've added all of the tests that I wanted to write.

gfyoung commented:
Thanks to whoever reran the build!  Everything is green now and ready for review.

chris-b1 commented on pandas/core/computation/eval.py:
```diff
@@ -148,7 +147,7 @@ def _check_for_locals(expr, stack_level, parser):
 def eval(expr, parser='pandas', engine=None, truediv=True,
          local_dict=None, global_dict=None, resolvers=(), level=0,
-         target=None, inplace=None):
+         target=None, inplace=False):
```
Defer to @jreback, @jorisvandenbossche, but I am still of the opinion that API would be cleaner if `inplace` was only allowed with a `target` and raised otherwise.  `pd.eval('1 + 2', inplace=True) == None` is surprising.

gfyoung replied:
@chris-b1 : True, but "with a `target`" is not well-defined either.  As you can see by my documentation change on `target` in `pandas.eval`, you first need to ensure that item assignment with strings is supported, but how do you check that without modifying the object?  That's something I imagined would create unnecessary overhead.  Also, you should then check if there is item assignment in the first place.

That's why I decided to allow the user to pass in `inplace=True` even if there is no real target.  I will in fact document that behavior just in case.

jorisvandenbossche replied:
As I said before, I don't think we need to check the modifyability of `target`. Leave it to the user.

gfyoung replied:
Yeah, sure, I have no problem leaving it as is.

jreback commented:
I don't really recall why there is a ``target`` object for assignment here at all. maybe @cpcloud remembers.

This seems odd to have ``target`` AND ``inplace``. Since inplace is our standard, would deprecate ``target``, default ``inplace=False``

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -76,6 +76,7 @@ Removal of prior version deprecations/changes
 
 - ``pd.read_excel()`` has dropped the ``has_index_names`` parameter (:issue:`10967`)
 - ``Categorical`` has dropped the ``.order()`` and ``.sort()`` methods in favor of ``.sort_values()`` (:issue:`12882`)
+- ``pd.eval`` and ``DataFrame.eval`` have changed the default of ``inplace`` from ``None`` to ``False`` (:issue:`11149`)
```
I like to use `` :func:`eval` `` so its a link (essentially we should always do this if feasible, maybe even make a mini-doc section on how to do this. e.g.  (OT). I think the `` :class:`~pandas.HDFStore` `` links are not working :<)

gfyoung replied:
Hmm...that's new (you could also make the argument to that for `pd.read_csv` for example).  What's the correct format:

``pd.eval`` --> :func:`eval`
``DataFrame.eval`` --> ???

jorisvandenbossche replied:
``` :method:`DataFrame.eval` ```

chris-b1 commented:
`target` is on `pd.eval` because `DataFrame.eval` calls it internally, and it is used to set the scope to calling df.

jreback commented:
> target is on pd.eval because DataFrame.eval calls it internally, and it is used to set the scope to calling df.

so then either pass a private arg ``_target``, then this is no problem. or do a wrapper method that ``DataFrame.eval`` calls (that handles the ``target`` stuff) rather than the public ``eval`` (maybe a bit more work, but prob 'cleaner')

gfyoung commented:
> so then either pass a private arg _target, then this is no problem. or do a wrapper method that DataFrame.eval calls

Ummm, `target` is pretty ingrained in the code.  I think the "cleanest" is to enforce `target` being a `DataFrame` (similar to what we do with `pd.merge`)

jreback commented:
@gfyoung not sure what you mean the ``target`` argument itself is almost the last kwarg.

gfyoung commented:
@jreback : I'm saying that we should change the signature to place the `target` as the first-class citizen in this function by putting it at the start of the signature just like we do in `pd.merge` for example.  That would help to clear up a lot of confusion regarding the specifications surrounding `pd.eval`.

In addition, we should require that `target` is a `DataFrame` (just like we do in `pd.merge`) as well as enforce that each line of the expressions have item assignment for this to make any sense (though I don't know how much `pd.eval` is used beyond this context of `DataFrame` item assignment).

jreback commented:
> @jreback : I'm saying that we should change the signature to place the target as the first-class citizen in this function by putting it at the start of the signature just like we do in pd.merge for example. That would help to clear up a lot of confusion regarding the specifications surrounding pd.eval.

seems rather extreme change for limited upside. 

> In addition, we should require that target is a DataFrame (just like we do in pd.merge) as well as enforce that each line of the expressions have item assignment for this to make any sense (though I don't know how much pd.eval is used beyond this context of DataFrame item assignment).

yes

gfyoung commented:
> seems rather extreme change for limited upside.

Really?  It's literally just moving the position of the parameter to indicate importance:

~~~
def eval(target, expr,...)
~~~

jorisvandenbossche commented:
> Really?

That breaks all user code that uses `eval`, so indeed rather invasive?

jorisvandenbossche commented:
I don't think we need to change anything about `target`. It's probably not something a lot of people use or find useful, but it is there for internal use and for the power user, so just lets leave it there.

gfyoung commented:
> That breaks all user code that uses eval, so indeed rather invasive?

Right, but as this is for a major version release, an API change like this isn't too drastic, is it?

> It's probably not something a lot of people use or find useful, but it is there for internal use and for the power user, so just lets leave it there.

That's fair.  As the implementation stands, I don't mind leaving it there.  However, I do think we will want to consider restricting it to just `DataFrame`, in which it should move up in the signature.

jorisvandenbossche commented:
> Right, but as this is for a major version release, an API change like this isn't too drastic, is it?

Well, even for a major release we don't just want to break something, it has to have a good reason. And I don't think there is a good reason here. And more importantly, it is even an optional argument, so IMO that makes no sense to put it first.

>  However, I do think we will want to consider restricting it to just DataFrame, in which it should move up in the signature.

Why is moving needed? I don't see any relation with restricting what you can pass to `target` and the need to move `target` in the signature

gfyoung commented:
So judging from the conversation, it seems like any drastic changes to `pd.eval` in terms of aligning it with other top-level functions `pd.merge` are probably not going to happen.

However, I'm okay with leaving the implementation as is (and not verify `target`).

jorisvandenbossche commented on pandas/core/computation/eval.py:
```diff
@@ -206,19 +205,26 @@ def eval(expr, parser='pandas', engine=None, truediv=True,
+    ValueError : 1) A non-None `target` was passed in that does not support
+                    string item assignment.
+                 2) `inplace=False`, and a non-None `target` was passed in
+                    that does not support `.copy()`.
```
This won't render nicely. I think you can put the list on the line below (in the 'description' part like for parameters)

gfyoung commented:
@jorisvandenbossche (in response to your comment) : I was proposing that we could drastically modify `pd.eval` to operate just on `DataFrame` (so require `target`), which is why we would move it up on the signature.  This is to align the behavior and signature with other top-level functions like `pd.merge`, which operate solely on `DataFrame`.

jorisvandenbossche commented:
I just don't see any argument for such a drastic change.

> I was proposing that we could drastically modify pd.eval to operate just on DataFrame (so require target), 

Ah, OK, then it makes some sense. But, for that you can already use `DataFrame.eval`. Further, that defeats the purpose of `eval` to just execute the expression without a dataframe as the context (it can be used for that, so probably people do it). 
It is true that now the equivalence between `pd.eval` and `DataFrame.eval` is not as consistent as eg with `merge` (pd.merge(df, ..) is indeed just df.merge(..)), but I don't think it is worth it for `eval` to change this.

gfyoung commented:
@jorisvandenbossche : Fair enough.  I think this PR does enough as it is, hence why I'm fine with leaving this discussion for another time.

jorisvandenbossche reviewed (commented):
I agree with @chris-b1 that it would be nicer to raise an error when `inplace` is set to True when no `target` is used / no assignment is in the expression.

When there is no assignment, the result should always be returned IMO

gfyoung commented:
> I agree with @chris-b1 that it would be nicer to raise an error when inplace is set to True when no target is used / no assignment is in the expression.

As I explained earlier, checking for all of those things is not straightforward to do and imposes unnecessary overhead.  The major reason is that "no target" used is not good enough.  You need to check that `target` does not support string item assignment.

Also, if you're going to trust the user regarding `target`, why are you checking `inplace` with respect to `target` ? That seems inconsistent with what you have said previously.

gfyoung commented:
I would be alright with raising if there is no assignment, but checking for `target` is something I would not do given the robustness issues.

jorisvandenbossche commented:
> As I explained earlier, checking for all of those things is not straightforward to do and imposes unnecessary overhead.

I don't see why this would give overhead. You already have all booleans you need (`target_modified` and `inplace`).

For example, could change the final return lines

```
if inplace is False:
    return target if target_modified else ret
```

with
```
if target_modified:
    if not inplace:
        return target
elif:
    if not inplace:
        return ret
    else:
        raise ValueError("inplace cannot be set to true when no target is being modified")
```

(didn't run, so could be some flaws in it) 
If we don't care about the error, you can just change the last 4 lines with `return ret`, so at least there is always a return value when no target/assignment is used.

> You need to check that target does not support string item assignment.

If this is the case, you got already an error before that?

> Also, if you're going to trust the user regarding target, why are you checking inplace with respect to target ? That seems inconsistent with what you have said previously.

Sorry, I fail to see the link :-) For the error OK, but at least if nothing is being modified, there should always be a return value. This would mean in practice that the `inplace` keyword is being ignored. Therefore I think it is nice to raise an error, but as I said above, also OK with just always returning the result.

gfyoung committed:
MAINT/BUG: Default inplace to False in pd.eval

Deprecated in 0.18.0.

xref gh-11149.

Also patches bug where we were improperly handling
the inplace=False condition, as we were assuming
that target input was non-None when that wasn't
necessarily enforced.

jorisvandenbossche commented on pandas/core/computation/eval.py:
```diff
@@ -206,19 +205,33 @@ def eval(expr, parser='pandas', engine=None, truediv=True,
+    ------
+    ValueError : There are many instances where such an error can be raised:
+        - An invalid target is provided, and
+          the expression is multiline.
```
Can you put this on one line? (I think it fits)

jorisvandenbossche replied:
Further, why the "and the expression is multiline" ?

gfyoung replied:
Meant to go for the condition of `multi_line and target=None`

jorisvandenbossche commented on pandas/core/computation/eval.py:
```diff
@@ -206,19 +205,33 @@ def eval(expr, parser='pandas', engine=None, truediv=True,
 
+    Raises
+    ------
+    ValueError : There are many instances where such an error can be raised:
```
Can you put the "There are many instances where such an error can be raised:" also on the next line? (and then the colon `:` is not needed)

gfyoung replied:
Done.

gfyoung commented on pandas/core/frame.py:
```diff
@@ -2224,21 +2224,18 @@ def query(self, expr, inplace=False, **kwargs):
-            explicitly rather than relying on the default.
+        inplace : bool, default False
+            If the expression contains an assignment, whether to perform the
+            operation inplace and mutate the existing DataFrame. Otherwise,
```
I'm not sure why we need to emphasize this.  It's pretty clear that the default is `False` from the declaration of `inplace` on line 2235.

gfyoung commented on pandas/tests/computation/test_eval.py:
```diff
@@ -1388,14 +1380,52 @@ def test_assignment_in_query(self):
+        expression = "a = 1 + 2"
+
+        with tm.assert_raises_regex(ValueError, msg):
+            self.eval(expression, target=invalid_target, inplace=False)
```
True, but I can also pass in objects that are constructed to support item assignment but fail on copy.  It seems reasonable to put a custom error message for that case.

gfyoung commented:
@jreback @jorisvandenbossche @chris-b1 : PTAL

gfyoung commented:
@jorisvandenbossche : Any updates or thoughts on this?

gfyoung commented:
@jreback @jorisvandenbossche @chris-b1 : Any updates or thoughts on this?

jreback commented on pandas/core/computation/eval.py:
```diff
@@ -206,19 +205,32 @@ def eval(expr, parser='pandas', engine=None, truediv=True,
         The number of prior stack frames to traverse and add to the current
         scope. Most users will **not** need to change this parameter.
     target : a target object for assignment, optional, default is None
```
this doc-string is not numpy doc (IOW the expl should be on the next line), specify *what* the target object can be (DataFrame?)

gfyoung replied:
It really can be anything that supports string item-assignment, per the docstring I put.

jreback commented on pandas/core/computation/eval.py:
```diff
@@ -206,19 +205,32 @@ def eval(expr, parser='pandas', engine=None, truediv=True,
-        WARNING: inplace=None currently falls back to to True, but
-        in a future version, will default to False.  Use inplace=True
-        explicitly rather than relying on the default.
+        This is essentially passed into the resolver and is used when there
```
rm essentially.

gfyoung replied:
Makes, sense.  Done.

jreback commented on pandas/core/computation/eval.py:
```diff
@@ -206,19 +205,32 @@ def eval(expr, parser='pandas', engine=None, truediv=True,
+        There are many instances where such an error can be raised:
+
+        - `target=None`, but the expression is multiline.
+        - The expression is multiline, but not all them have item assignment.
```
expand what this means (for 2nd item)

gfyoung replied:
Sure, done.

jreback commented on pandas/core/computation/eval.py:
```diff
@@ -266,28 +282,33 @@ def eval(expr, parser='pandas', engine=None, truediv=True,
+                try:
+                    target = env.target.copy()
+                except AttributeError:
+                    raise ValueError("Cannot return a copy of the target")
```
do we have a test for this?

gfyoung replied:
Absolutely.  There is an entire test devoted to this!

jreback reviewed (approved):
just some minor comments.

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -60,6 +60,7 @@ Other API Changes
 ^^^^^^^^^^^^^^^^^
 
 - Moved definition of ``MergeError`` to the ``pandas.errors`` module.
+- :func:`eval` will now raise a ``ValueError`` in cases where we cannot operate properly on the ``target`` parameter, or when an inplace operation is specified but there is no item assignment in the expression (:issue:`16732`)
```
I think it makes sense to have a couple of examples in a sub-section on what is happening. Also feel free to expand the Examples section of the doc-string.

gfyoung replied:
I tried to clarify the wording a bit.  Done.

gfyoung commented:
@jreback @jorisvandenbossche @chris-b1 : Addressed all comments.  PTAL.

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -60,6 +60,7 @@ Other API Changes
 ^^^^^^^^^^^^^^^^^
 
 - Moved definition of ``MergeError`` to the ``pandas.errors`` module.
+- :func:`eval` will now raise a ``ValueError`` when item assignment malfunctions, or inplace operations are specified, but there is no item assignment in the expression (:issue:`16732`)
```
can you add a small sub-section with an example of what is changing

gfyoung replied:
Done.

gfyoung replied:
Added it to the breaking section, since technically, passing in `inplace=True` without item assignment is disallowed now.

jreback replied:
nice can u make title shorter

gfyoung commented:
@jreback @jorisvandenbossche @chris-b1 : Addressed comment.  PTAL.

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -45,6 +45,48 @@ Other Enhancements
+change, the error message is now this:
+
+.. ipython :: python
+
```
i think u need an okexception or something here
(or a code block)

gfyoung replied:
Used code block.

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -45,6 +45,48 @@ Other Enhancements
+
+.. ipython :: python
+
+   pd.eval("1 + 2", target=arr, inplace=True)
```
same

gfyoung replied:
Done.

gfyoung commented:
@jreback @jorisvandenbossche @chris-b1 : Addressed comments.  PTAL.

gfyoung commented:
Also, just a reminder, xref #6581 (deprecation log) and #11149 (PR with deprecation)

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -45,6 +45,52 @@ Other Enhancements
+  IndexError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`)
+  and integer or boolean arrays are valid indices
+
+This is a very long way of saying numpy arrays don't support string-item indexing. With this
```
don't -> doesn't

gfyoung replied:
It's plural ("arrays"), so "don't" is the correct conjugation there.

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -45,6 +45,52 @@ Other Enhancements
+  Out[4]: 3
+
+However, this input does not make much sense because the output is not being assigned to
+the target. Now, a `ValueError` will be raised when such an input is passed in:
```
double-backticks

gfyoung replied:
Done.

jreback reviewed (approved):
some tiny doc comments. ping when pushed.

gfyoung committed:
DOC: Double-tick ValueError in whatsnew

jreback commented:
thanks @gfyoung
